pkg/apis/apps/v1alpha1/types.go

```diff
@@ -98,6 +99,319 @@ type ApplicationSyncPolicy struct {
+	// The header keys must be lowercase and use hyphen as the separator.
+	// values are case-sensitive and formatted as follows:
+	// - `exact: "value"` for exact string match
+	// - `prefix: "value"` for prefix-based match
+	// - `regex: "value"` for ECMAscript style regex-based match
+	// e.g.:
+	// match:
+	//   - headers:
+	//       myheader:
+	//         regex: ".*XXXX.*"
+	//   - headers:
+	//       cookie:
+	//         regex: "^(.*?;)?(type=insider)(;.*)?$"
+	// Note: If you want to use A/B Testing, you need to configure analysisTimes and match.
+	// If you only configure analysisTimes, it will trigger Blue/Green Deployment.
+	// You can configure both canaryStrategy and match.
+	// If configure both canaryStrategy and match, Traffic that meets match goes towards the preview service.
+	// Traffic that doesn't meet the match will go to the primary service and preview service proportionally.
+	// +optional
+	Match []istiov1alpha3.HTTPMatchRequest `json:"match,omitempty"`
+}
+
+type CanaryConfig struct {
+	// Max traffic weight routed to canary test
+	// If empty and no stepweights are set, 100 will be used by default.
+	// +optional
+	MaxWeight int `json:"maxWeight,omitempty"`
+
+	// StepWeight defines the incremental traffic weight step for analysis phase
+	// If set stepWeight: 10 and set maxWeight: 50
+	// The flow ratio between PREVIEW and PRIMARY at each step is
+	// (10:90) (20:80) (30:70) (40:60) (50:50)
+	// +optional
+	StepWeight int `json:"stepWeight,omitempty"`
+
+	// StepWeights defines the incremental traffic weight steps for analysis phase
+	// Note: Cannot configure stepWeights and stepWeight at the same time.
+	// If both stepWeights and maxWeight are configured, the traffic
+	// will be scaled according to the settings in stepWeights only.
+	// If set stepWeights: [1, 10, 20, 80]
+	// The flow ratio between PREVIEW and PRIMARY at each step is
+	// (1:99) (10:90) (20:80) (80:20)
+	// +optional
+	StepWeights []int `json:"stepWeights,omitempty"`
+
+	// StepWeightPromotion defines the incremental traffic weight step for promotion phase
+	// If maxWeight: 50 and set stepWeightPromotion: 20
+	// After a successful test, traffic to the PRIMARY version changes as follows: 50 70 90 100.
+	// +optional
+	StepWeightPromotion int `json:"stepWeightPromotion,omitempty"`
+}
+
+type TrafficAnalysis struct {
+	// CheckIntervalSeconds defines the schedule interval for this traffic analysis.
+	// Interval is the time interval between each test.
+	// Kurator changes the traffic distribution rules (if they need to be changed)
+	// and performs a traffic analysis every so often.
+	// Defaults to 60.
+	CheckIntervalSeconds *int `json:"checkIntervalSeconds"`
+
+	// CheckFailedTimes defines the max number of failed checks before the traffic analysis is terminated
+	// If set "checkFailedTimes: 2". It means Kurator will rollback when check failed 2 times.
+	CheckFailedTimes *int `json:"checkFailedTimes"`
```

Comment:
is there some limitation, is 0 a valid value?

Reply:
0 is a vaild value.